### PR TITLE
docs(gatsby-plugin-sharp): Base64 features

### LINKED ISSUE #25640: Gatsby Image & Sharp `background` blur-up
## Description

When using `fit: CONTAIN background: "#ffffff"` on a `childImageSharp` with the `...GatsbyImageSharpFluid_withWebp` fragment the initial blurred base64 image has a black background which changes to the white background with the image loaded.

### Steps to reproduce

Using a GraphQL static query, add 

```graphql
childImageSharp {
  fluid(
    maxWidth: 512
    maxHeight: 512
    fit: CONTAIN
    background: "#ffffff"
  ) {
    ...GatsbyImageSharpFluid_withWebp
  }
}
```

query to your image, 

see image blur up from black image to white.

Screen Recording 2020-07-09 at 12 27 37 PM

CodeSandbox Example

https://codesandbox.io/s/brave-chaplygin-p7tnw?file=/src/components/image.js

### Expected result

Blurred up image should have a white background

### Actual result

Blurred up image has a black background in Base64 image

### Environment
  System:
    OS: macOS 10.15.5
    CPU: (8) x64 Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 14.5.0 - /usr/local/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.14.5 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 83.0.4103.116
    Safari: 13.1.1
  npmPackages:
    gatsby: ^2.23.12 => 2.23.12 
    gatsby-image: ^2.4.9 => 2.4.9 
    gatsby-plugin-manifest: ^2.4.14 => 2.4.14 
    gatsby-plugin-postcss: ^2.3.9 => 2.3.9 
    gatsby-plugin-react-helmet: ^3.3.6 => 3.3.6 
    gatsby-plugin-sharp: ^2.6.14 => 2.6.14 
    gatsby-remark-images: ^3.3.17 => 3.3.17 
    gatsby-remark-normalize-paths: ^1.0.0 => 1.0.0 
    gatsby-remark-relative-images: ^0.3.0 => 0.3.0 
    gatsby-source-filesystem: ^2.3.14 => 2.3.14 
    gatsby-transformer-remark: ^2.8.23 => 2.8.23 
    gatsby-transformer-sharp: ^2.5.7 => 2.5.7

### PR DESCRIPTION
**Pull request by polarathene**
## Description

Documents some base64 features that Gatsby has [supported for a while now](https://github.com/gatsbyjs/gatsby/pull/10944). These weren't easily discoverable or clear how to use them.

I did some minor refactoring related to these features. The relevant individual commits provide details in their commit messages.

### Documentation

The graphql parameters `toFormatBase64` and `base64Width` are now documented in the `gatsby-plugin-sharp` README. The global override equivalent options are mentioned for each, but not documented elsewhere.

@gatsbyjs/learning is there any additional documentation of these that would be appreciated? A code snippet or PR to a project that demonstrates the minor features?

### Improvements

Some code refactoring, fixes and tests added.

Also `background` option is now passed to `base64()`, fixes #25640

### CODE REVIEW

File: packages/gatsby-plugin-sharp/README.md [hunk 1]
```diff
@@ -30,9 +30,13 @@ plugins: [
...
     resolve: `gatsby-plugin-sharp`,
     options: {
-      useMozJpeg: false,
+      // Available options and their defaults:
+      base64Width: 20,
+      forceBase64Format: ``, // valid formats: png,jpg,webp
+      useMozJpeg: process.env.GATSBY_JPEG_ENCODER === `MOZJPEG`,
       stripMetadata: true,
-      defaultQuality: 75,
+      lazyImageGeneration: true,
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/README.md:38]: Currently, this argument doesn't do anything. So we shouldn't display it in my opinion.
```suggestion
```
```

**Comment by LekoArts** [comment]
Seems like tests are failing cc @wardpeet @polarathene

**Comment by polarathene** [comment]
**build-test ** failure seems unrelated?

Unit tests is a mock snapshot that needs to be updated, I'll take care of that.

**Comment by polarathene** [comment]
I'm not familiar with the testing code. When sharp pipeline calls `resize()`, many tests are failing with the error:

```
Expected positive integer for height but received NaN of type number
```

I am assuming this is because of `base64Width` being moved out from a function call(`defaultBase64Width()`) which provided the default value, to now depending on the default value from `pluginDefaults` object which now lives in `plugin-options.js`. Should the tests be including those defaults?

**Comment by polarathene** [comment]
I'm unable to get a working test environment locally for `gatsby-plugin-sharp`, jest raises failures about finding modules that should be available but it can't find them.

If someone has time to assist that would be helpful.Otherwise not much else I can do with this PR.

**Comment by polarathene** [comment]
All tests should be passing, "Cloud Tests" seems to be hanging for some reason.

**Comment by polarathene** [comment]
Added a fix for `background` option not being passed from fluid/fixed methods to the base64 method.

Included a basic test , although like the base64 tests I've added I'm not sure if they're sufficient, should I also be testing these via fluid/fixed tests instead of directly on base64? The bug for `background` was simply that the option wasn't passed on to the method, is it worth adding tests for that? 

I'm not sure if I can mock the base64 method called in that test file for a single test that ensures supported options are passed through?

**Comment by polarathene** [comment]
@wardpeet Could you review/merge please?

**Comment by polarathene** [comment]
@wardpeet @pvdz hello, friendly reminder request to take the next steps in resolving this PR, thank you.

File: packages/gatsby-plugin-sharp/README.md [hunk 2]
```diff
@@ -147,6 +152,26 @@ following:
 Convert the source image to one of the following available options: `NO_CHANGE`,
 `JPG`, `PNG`, `WEBP`.
 
+#### toFormatBase64
+
+base64 image uses the image format from the source , or the value of `toFormat`. This setting allows a different image format instead, available options are: `JPG`, `PNG`, `WEBP`.
```
Comment (muescha) [inline packages/gatsby-plugin-sharp/README.md:157]: in code they are lowercased and there is no case insensitive comparison is used

File: packages/gatsby-plugin-sharp/src/index.js [hunk 3]
```diff
@@ -323,16 +323,16 @@ async function generateBase64({ file, args = {}, reporter }) {
     .png({
       compressionLevel: options.pngCompressionLevel,
       adaptiveFiltering: false,
-      force: args.toFormat === `png`,
+      force: options.toFormat === `png`,
```
Comment (muescha) [inline packages/gatsby-plugin-sharp/src/index.js:326]: better compare case insensitive?

Comment (polarathene) [reply packages/gatsby-plugin-sharp/src/index.js:326]: It's handled earlier in the `healOptions()` call:

https://github.com/gatsbyjs/gatsby/blob/1419afe21a48fc4d75909e054760693fadc8675f/packages/gatsby-plugin-sharp/src/plugin-options.js#L83-L84

File: packages/gatsby-plugin-sharp/README.md (hunk 2 above)
Comment (polarathene) [reply packages/gatsby-plugin-sharp/README.md:157]: It uses uppercase like the docs above it for `toFormat`. It calls `healOptions()` to lowercase just like `toFormat` does.

**Comment by polarathene** [comment]
> Added one question, looks good besides that.

@wardpeet where is that question?

File: packages/gatsby-plugin-sharp/src/plugin-options.js [hunk 4]
```diff
@@ -2,7 +2,8 @@ const _ = require(`lodash`)
 
 /// Plugin options are loaded onPreBootstrap in gatsby-node
 const pluginDefaults = {
-  forceBase64Format: false,
+  base64Width: 20,
+  forceBase64Format: ``, // valid formats: png,jpg,webp
```
Comment (wardpeet) [inline packages/gatsby-plugin-sharp/src/plugin-options.js:6]: This got omitted somehow, I probably didn't click on start-review...
Can't we change this to undefined instead? Technically this becomes an enum with 3 values then. We could add validation later on.
```suggestion
```

Comment (polarathene) [reply packages/gatsby-plugin-sharp/src/plugin-options.js:6]: @wardpeet You want to change the value of empty string to `undefined`? I've just matched the value to what is already used for `toFormat`

https://github.com/gatsbyjs/gatsby/blob/1419afe21a48fc4d75909e054760693fadc8675f/packages/gatsby-plugin-sharp/README.md#L142-L143

https://github.com/gatsbyjs/gatsby/blob/1419afe21a48fc4d75909e054760693fadc8675f/packages/gatsby-plugin-sharp/src/plugin-options.js#L26-L27

`toFormatBase64` like `toFormat` is the per image option, while `forceBase64Format` is a global default setting (perhaps the name doesn't communicate that well? It's in `pluginDefaults` object, so should hopefully be clear that it's forcing the default format for base64?)

Comment (wardpeet) [reply packages/gatsby-plugin-sharp/src/plugin-options.js:6]: That's what I was missing :) Thank you for the explanation! 👍 🌮

**Pull request merged**